packages/e2e-tests/specs/change-detection.test.js

```diff
@@ -151,6 +151,7 @@ describe( 'Change detection', () => {
 
 	it( 'Should prompt if content added without save', async () => {
 		await clickBlockAppender();
+		await page.keyboard.type( 'Paragraph' );
```

Comment:
This is changing the meaning of the test. The test was saying that if you just click the appender, you don't have to type to consider the post dirty.

Reply:
Yeah, but an empty paragraph block does not make the post dirty anymore. Other blocks do though.

Reply:
Why? why it was changed?

Reply:
Because:

```
// A single unmodified default block is assumed to
// be equivalent to an empty post.
if (
	blocksForSerialization.length === 1 &&
	isUnmodifiedDefaultBlock( blocksForSerialization[ 0 ] )
) {
	blocksForSerialization = [];
}
```

Reply:
Still, why it wasn't working this way before? Why are we making this User Experience change in this refactoring PR?

Reply:
Because, before, `blocks` were checked for referential equality in the editor to see if it was dirty or not. A new empty array was a dirty change to the editor. Now, we care about the content, and setting the content of an empty post to an empty string is not a dirtying change.

It doesn't affect users negatively does it? If anything, this behavior is technically the "correct" one.

Reply:
Thanks for the explanation and I can live with that change I think but I'm really concerned about the way we check dirtiness now in terms of performance. This forces us to serialize the blocks on each change which is not performant at all, especially for long posts.

Can't we rely on the presence of "edits" to check for dirtiness (like before).

Reply:
Yeah, that's what we do.

But:

```js
// Clear edits when they are equal to their persisted counterparts
// so that the property is not considered dirty.
```

We need that so that when we undo a post back to its original state, it's not dirty anymore.

Reply:
While I can see some value in this change. I wonder why it's in this PR. Why can't we keep the previous behavior because:

 - There are real performance concerns with this new behavior
 - It should be evaluated in its own PR whether this way of computing dirtiness is better than the previous one.

Reply:
This isn't what is forcing us to serialize blocks. We serialize blocks because `.content` expects a string, not an array of blocks.

Serialization should be even faster than rendering the block list, because `save` components are generally simpler than `edit` components, and we are not manipulating a DOM.

There is always the possibility of tagging `blocks` in `core` in a way that tells it to do something with it when saving, that something being serializing it and setting `content` to the result. But, it feels very hacky and we would also need to do it when comparing edits with persisted values.

Reply:
It was an early decision to not serialize blocks on each change because it's too expensive. (You can probably verify quickly by running the performance tests on this branch).

`edit` is not expensive because React doesn't rerender everything on each change and with our asyncmode we only rerender synchronously the edited block.

Reply:
(It only happens on every persistent/undo-level-creating change.)

Reply:
> This branch

Average time to type character: 78.965ms
Slowest time to type character: 170ms
Fastest time to type character: 69ms

> master

Average time to type character: 52.535ms
Slowest time to type character: 122ms
Fastest time to type character: 43ms

Reply:
https://github.com/WordPress/gutenberg/pull/16932/commits/1f4889721a4c4cac0a8ac863acef5bb24242b3ae